### PR TITLE
[WIP] Directory

### PR DESCRIPTION
**Pull request by garrensmith**
This is the PR to get the directory layer working with the rust client. Its currently still very rough
and doesn't even compile yet.

If any of the maintainers want to contribute feel free to push code. I'm also on the rust discord (redcomet) if you want to chat about it.

### CODE REVIEW

**Comment by garrensmith** [comment]
Just to add I've based the initial work off the golang and python implementation. I've also used the [notes](https://github.com/cloudant-labs/couchdb-erlfdb/blob/master/notes/directory-layer-design.md) from the erlang implementation that we using for CouchDB.

File: scripts/run_bindingtester.sh [hunk 1]
```diff
@@ -6,7 +6,7 @@ fdb_rs_dir=$(pwd)
 
 case $(uname) in
   Darwin)
-    brew install mono
+#    brew install mono
```
Comment (Speedy37) [inline scripts/run_bindingtester.sh]:
```suggestion
command -v mono || brew install mono
```

File: scripts/run_bindingtester.sh [hunk 2]
```diff
@@ -22,7 +22,7 @@ esac
   cd ${fdb_builddir:?}
 
   ## Get foundationdb source
-  git clone --depth 1 https://github.com/apple/foundationdb.git -b release-6.1
+#  git clone --depth 1 https://github.com/apple/foundationdb.git -b release-6.1
```
Comment (Speedy37) [inline scripts/run_bindingtester.sh]:
```suggestion
[[ -d foundationdb ]] || git clone --depth 1 https://github.com/apple/foundationdb.git -b release-6.1
```

File: foundationdb/src/directory/node.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+        return self;
+    }
+
+    pub async fn layer(&mut self, trx: &Transaction) -> &[u8] {
+        if self._layer == None {
+            let key = self.subspace.subspace(&b"layer".to_vec());
+            self._layer = match trx.get(key.bytes(), false).await {
+                Ok(None) => Some(Vec::new()),
+                Err(_) => Some(Vec::new()),
```
Comment (Speedy37) [inline foundationdb/src/directory/node.rs]: I don't think silencing an FdbError is ok.
layer should probably return a Result

File: foundationdb/src/directory/node.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+use crate::future::FdbSlice;
+use crate::{Transaction, Directory, DirectoryResult};
+use crate::directory::directory_subspace::DirectorySubspaceResult;
+
+
+pub struct Node {
+    subspace: Subspace,
+    path: Vec<String>,
+    target_path: Vec<String>,
+    _layer: Option<Vec<u8>>
```
Comment (Speedy37) [inline foundationdb/src/directory/node.rs]: Layer being used as a cache you might want to consider https://docs.rs/futures/0.3.4/futures/future/trait.FutureExt.html#method.shared.
RefCell might also be required in order to not requires `&mut self` on a lot of methods (but be careful about dynamic borrows lifespans in async context).

**Pull request closed without merging**